src/services/storage/Storage.ts

```diff
@@ -0,0 +1,184 @@
+import { isDefined } from '@prefecthq/prefect-design'
+import { ComputedRef, Ref, computed, reactive, ref, unref, watch, watchEffect } from 'vue'
+import { StorageItem } from '@/services/storage/StorageItem'
+
+type Store<T extends StorageItem> = Map<string, T | undefined>
+type Subscriptions = Map<string, Set<string> | undefined>
+
+type StorageSubscription<T extends StorageItem | StorageItem[]> = {
+  value: ComputedRef<T | undefined>,
+  unsubscribe: Unsubscribe,
+}
+
+export type Unsubscribe = () => void
+
+function waitForCondition<T>(input: Ref<T>, condition: (value: T) => boolean): Promise<void> {
+  return new Promise(resolve => {
+    watchEffect(() => {
+      if (condition(input.value)) {
+        resolve()
+      }
+    })
+  })
+}
+
+export class Storage<T extends StorageItem> {
+  private readonly subscriptions: Subscriptions = new Map()
+  private readonly store: Store<T> = reactive(new Map())
+
+  public add(input: T | Ref<T | undefined>): void {
+    const inputRef = ref(input) as Ref<T | undefined>
+    const value = unref(inputRef)
+
+    if (value === undefined) {
+      this.addWhenDefined(inputRef)
+      return
+    }
+
+    if (this.isNewValue(value)) {
+      this.store.set(value.id, value)
+    }
+  }
+
+  private async addWhenDefined(input: Ref<T | undefined>): Promise<void> {
+    await waitForCondition(input, isDefined)
+
+    this.add(input)
+  }
+
+  public remove(id: string): void {
+    this.subscriptions.delete(id)
+    this.store.delete(id)
+  }
+
+  public addAll(inputs: T[] | Ref<T[] | undefined>): void {
+    const inputsRef = ref(inputs) as Ref<T[] | undefined>
+    const values = unref(inputsRef)
+
+    if (values === undefined) {
+      this.addAllWhenDefined(inputsRef)
+      return
+    }
+
+    values.forEach(value => this.add(value))
+  }
+
+  private async addAllWhenDefined(input: Ref<T[] | undefined>): Promise<void> {
+    await waitForCondition(input, isDefined)
+
+    this.addAll(input)
+  }
+
+  public removeAll(ids: string[]): void {
+    ids.forEach(id => this.remove(id))
+  }
+
+  public subscribe(id: string | Ref<string | null | undefined>): StorageSubscription<T> {
+    const idRef = ref(id)
+
+    const value = computed(() => {
+      if (idRef.value) {
+        return this.store.get(idRef.value)
+      }
+
+      return undefined
+    })
+
+
+    let unsubscribe: Unsubscribe | undefined = undefined
+
+    watch(idRef, (id) => {
```

Comment:
Can't totally tell if this watcher get's destroyed when the `unsubscribe` is called?

Reply:
I ended up removing this